File: day08/playground.py
import sys
from itertools import combinations
import networkx as nx
from math import prod

def dist(b1, b2):
    return sum([(b1[0]-b2[0])**2, (b1[1]-b2[1])**2, (b1[2]-b2[2])**2])

def main():
    fname = sys.argv[1]
    num_juncs = int(sys.argv[2])
    with open(fname) as f:
        boxes = []
        for line in f:
            p1,p2,p3 = [int(n) for n in line.split(',')]
            boxes.append((p1,p2,p3))
                        

    G = nx.Graph()
    G.add_nodes_from(boxes)
        
    dists = {(b1,b2): dist(b1, b2) for b1, b2 in combinations(boxes, 2)}

    sorted_juncs = [item[0] for item in sorted(dists.items(), key=lambda item: item[1])]
    for box_pair in sorted_juncs[:num_juncs]:
        b1, b2 = box_pair
        G.add_edge(b1, b2)

    print('Part 1:', prod([len(c) for c in sorted(nx.connected_components(G), key=len, reverse=True)][0:3]))
    
    i = num_juncs
    while nx.number_connected_components(G) > 1:
        b1, b2 = sorted_juncs[i]
        G.add_edge(b1, b2)
        i += 1

    print('Part 2:', b1[0] * b2[0])

main()

File: day08/playground.pl
#!/usr/bin/env perl
use v5.42;
use autodie;
use Graph::Undirected;
use List::Util qw(sum);

sub dist($b1, $b2) {
    return sum( map { ($b1->[$_] - $b2->[$_]) ** 2 } 0..2);
}

my ($fname, $num_juncs) = @ARGV;
open my $fh, '<', $fname;
my @boxes;
while (<$fh>) {
    chomp;
    push @boxes, [split ','];
}

my $g = Graph::Undirected->new;
for my $box (@boxes) {
    $g->add_vertex("@$box");
}

my %dists;
for my $i (0..$#boxes-1) {
    my $b1 = $boxes[$i];
    for my $j ($i+1..$#boxes) {
        my $b2 = $boxes[$j];
        $dists{"@$b1,@$b2"} = dist($b1, $b2);
    }
}

my @sorted = sort {$dists{$a} <=> $dists{$b}} keys %dists;
for my $i (0..$num_juncs-1) {
    my ($b1, $b2) = split ',', $sorted[$i];
    $g->add_edge($b1, $b2);
}
my @cc = $g->connected_components;
my @sorted_cc = sort { @$b <=> @$a } @cc;
my $part1 = 1;
for my $i (0..2) {
    $part1 *= @{$sorted_cc[$i]};
}

say "Part 1: $part1";

my $i = $num_juncs;
my ($b1, $b2);
while (scalar $g->connected_components > 1) {
    ($b1, $b2) = split ',', $sorted[$i];
    $g->add_edge($b1, $b2);
    $i++;
}

my @b1 = split ' ', $b1;
my @b2 = split ' ', $b2;
say "Part 2: ", $b1[0] * $b2[0];
